README.md
# Nuvla deployment on Kubernetes

[![BUILD](https://github.com/nuvla/deployment-k8s/actions/workflows/main.yml/badge.svg)](https://github.com/nuvla/deployment-k8s/actions/workflows/main.yml)

This repository contains the definitions for the deployment of the Nuvla
platform on Kubernetes cluster.

 - [demo](https://github.com/nuvla/deployment-k8s/tree/master/demo): contains
 deployment definitions for quick provisioning of Nuvla for demo or 
 evaluation. The deployment uses the latest released components of Nuvla. 
 Data is not persisted across restarts of the Nuvla DB layer.
 
 - [test](https://github.com/nuvla/deployment-k8s/tree/master/test): contains
 deployment definitions for quick provisioning of Nuvla for testing or
 evaluation. The deployment is done using latest Nuvla artifacts from the master
 branch. Data is not persisted across restarts of the Nuvla DB layer.
 
 - [prod](https://github.com/nuvla/deployment-k8s/tree/master/prod): contains
 deployment definitions suitable for deploying Nuvla for production usage.

# Contributing

### Source Code Changes

To contribute code to this repository, please follow these steps:

 1. Create a branch from master with a descriptive, kebab-cased name
    to hold all your changes.

 2. Follow the developer guidelines concerning formatting, etc. when
    modifying the code.
   
 3. Once the changes are ready to be reviewed, create a GitHub pull
    request.  With the pull request, provide a description of the
    changes and links to any relevant issues (in this repository or
    others). 
   
 4. Ensure that the triggered CI checks all pass.  These are triggered
    automatically with the results shown directly in the pull request.

 5. Once the checks pass, assign the pull request to the repository
    coordinator (who may then assign it to someone else).

 6. Interact with the reviewer to address any comments.

When the reviewer is happy with the pull request, he/she will "squash
& merge" the pull request and delete the corresponding branch.

### Testing

Add appropriate tests that verify the changes or additions you make to
the source code.

### Code Formatting

This repository contains mostly Docker container descriptions and bash
scripts. When modifying a file, keep the style of the existing code.

## Copyright

Copyright &copy; 2021, SixSq SA

## License

Licensed under the Apache License, Version 2.0 (the "License"); you
may not use this file except in compliance with the License.  You may
obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
implied.  See the License for the specific language governing
permissions and limitations under the License.


CHANGELOG.md
# Changelog

## 2021-11-03

# Changed

  - fixed running e2e tests via GHA
  - added run of tests via test2junit with xml reports
  - added publication of test results via GHA
  - added GHA to release the demo chart on github and publish it via GH pages
  - updated documentation
  - restructured the project 

## 2021-11-01

# Added
   - Initial deployment of Nuvla with Helm. 


test/nuvla-delete.sh
#!/bin/bash

kubectl delete -f namespace.yaml
kubectl delete -f traefik-rbac.yaml


test/nuvla-deploy.sh
#!/bin/bash

set -e

kubectl create -f namespace.yaml

kubectl -n nuvla-test create -f zk-deployment.yaml

kubectl -n nuvla-test create -f es-deployment.yaml

kubectl -n nuvla-test create -f api-deployment.yaml

kubectl -n nuvla-test create -f jobs-deployment.yaml

kubectl -n nuvla-test create -f ui-deployment.yaml

kubectl create -f traefik-rbac.yaml
kubectl -n nuvla-test create configmap traefik-conf --from-file=traefik.toml
kubectl -n nuvla-test create -f traefik-deployment.yaml
kubectl -n nuvla-test create -f traefik-ingress.yaml


test/swarm-deploy.sh
#!/bin/bash
set -e

cleanup() {
    $DM_BIN rm -y "$($DM_BIN ls -q)"
}

trap cleanup ERR

action_err_msg="Action required: deploy|terminate"
ACTION=${1:?$action_err_msg}
SWSIZE=${2:-1}

DM_VER=v0.16.2
DM_BIN=$HOME/docker-machine

if [ ! -f "$DM_BIN" ]; then
    base=https://github.com/docker/machine/releases/download/$DM_VER
    curl -L $base/docker-machine-"$(uname -s)"-"$(uname -m)" >"$DM_BIN"
fi
chmod +x "$DM_BIN"

if [ ! -f "$HOME"/.ssh/id_rsa ]; then
    yes y | ssh-keygen -q -t rsa -N '' -f "$HOME"/.ssh/id_rsa &>/dev/null
fi

MNAME=dockermachine-$(date +%s)
CREATE_CMD="create --driver exoscale 
    --exoscale-ssh-user root 
    --exoscale-ssh-key $HOME/.ssh/id_rsa
    --exoscale-api-key ${EXOSCALE_API_KEY} 
    --exoscale-api-secret-key ${EXOSCALE_API_SECRET} 
    --exoscale-availability-zone CH-GVA-2 
    --exoscale-instance-profile Large 
    --exoscale-security-group slipstream_managed"

deploy() {
    swsize=${1:-1}
    echo "::: Provisioning master: $MNAME"
    $DM_BIN $CREATE_CMD --exoscale-image 'Linux Ubuntu 18.04 LTS 64-bit' "$MNAME"
    ip=$($DM_BIN ip "$MNAME")
    $DM_BIN ssh "$MNAME" "sudo docker swarm init --force-new-cluster --advertise-addr $ip"
    
    if [ "$swsize" -gt 1 ]; then
        joinToken=$($DM_BIN ssh "$MNAME" "sudo docker swarm join-token worker -q" | tr -d '\r')
        for i in $(seq 1 $((swsize - 1))); do
            WNAME=${MNAME}-worker${i}
            echo "::: Provisioning worker: $WNAME"
            $DM_BIN $CREATE_CMD --exoscale-image 'Linux Ubuntu 18.04 LTS 64-bit' "$WNAME"
            $DM_BIN ssh "$WNAME" "sudo docker swarm join --token ${joinToken} ${ip}:2377"
        done
    fi

    $DM_BIN ssh "$MNAME" "sudo docker node ls"

    echo "$MNAME" > "$HOME"/docker-machine-master
}

terminate() {
    machines=()
    while read m;do machines+=( "$m" );done < <($DM_BIN ls -q)
    if [ "${#machines[@]}" -eq 0 ]; then
        echo "WARNING: no machines to terminate"
    else
        for m in ${machines[@]};do
            $DM_BIN rm -y "$m" &
        done
        wait
    fi
}

if [ "$ACTION" == "deploy" ]; then
    deploy "$SWSIZE"
elif [ "$ACTION" == "terminate" ]; then
    terminate
else
    echo "$action_err_msg"
    exit 1
fi


k8s-cluster/k8s-install.sh
#!/bin/bash
set -ex

ROLE=${1:?"Role: master or worker"}

K8S_VER=${2:-1.19.5-00}

# "Fix" docker daemon: https://github.com/kubernetes/kubeadm/issues/1394#issuecomment-462878219
cat > /tmp/daemon.json <<EOF
{
  "exec-opts": ["native.cgroupdriver=systemd"],
  "log-driver": "json-file",
  "log-opts": {
    "max-size": "100m"
  }
}
EOF
sudo mv /tmp/daemon.json /etc/docker/daemon.json
sudo mkdir -p /etc/systemd/system/docker.service.d
sudo systemctl daemon-reload
sudo systemctl restart docker

sudo sh -c "echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections"
sudo apt-get update
sudo apt-get install dialog apt-utils -y -q

sudo apt-get install -y apt-transport-https
sudo curl https://packages.cloud.google.com/apt/doc/apt-key.gpg --output /etc/apt/trusted.gpg.d/k8s-apt-key.gpg
sudo apt-add-repository "deb http://apt.kubernetes.io/ kubernetes-xenial main" -y
sudo apt update
sudo apt-get install -y kubelet=$K8S_VER kubeadm=$K8S_VER kubectl=$K8S_VER

if [ "$ROLE" == "master" ]; then
    sudo apt-get install -y --allow-downgrades kubectl=$K8S_VER
    sudo kubeadm init --pod-network-cidr=10.244.0.0/16
    mkdir $HOME/.kube
    sudo cp /etc/kubernetes/admin.conf /home/ubuntu/.kube/config
    sudo chown ubuntu. .kube/config
    kubectl cluster-info
    KN=https://raw.githubusercontent.com/coreos/flannel/master/Documentation/kube-flannel.yml
    kubectl apply -f $KN | tee pod_network_setup.txt
    kubectl taint nodes --all node-role.kubernetes.io/master-
    kubectl get nodes -o wide
    sudo sed -i -e '/service-cluster-ip-range/a\' -e '    - --service-node-port-range=80-9000' /etc/kubernetes/manifests/kube-apiserver.yaml
fi


functional-tests/project.clj
(def +sixsq-nuvla-api-version+ "2.0.0")

(defproject sixsq.nuvla.deployment/functional-tests "0.0.1-SNAPSHOT"

  :description "test full deployment with clojure library"

  :url "https://github.com/nuvla/deployment"

  :license {:name         "Apache 2.0"
            :url          "http://www.apache.org/licenses/LICENSE-2.0.txt"
            :distribution :repo}

  :plugins [[lein-parent "0.3.5"]]

  :parent-project {:coords  [sixsq.nuvla/parent "6.5.0"]
                   :inherit [:plugins
                             :min-lein-version
                             :managed-dependencies
                             :repositories
                             :deploy-repositories]}

  :pom-location "target/"

  :clean-targets ^{:protect false} ["target" "out"]

  :dependencies
  [[sixsq.nuvla/api ~+sixsq-nuvla-api-version+]
   [environ]
   [clj-http]]

  :profiles {:provided {:dependencies [[org.clojure/clojure]]}
             :dev {:plugins [[test2junit "1.4.2"]]}
             :test     {:source-paths   ["test"]
                        :resource-paths ["dev-resources"]}})


k8s-cluster/k8s-deploy-exoscale.sh
#!/bin/bash
set -xe

source ./env.sh

action_err_msg="Action required: deploy|terminate"
ACTION=${1:?$action_err_msg}
SWARM_NODE_COUNT=${2:-1}
WORKER_VM_PROFILE=${3:-'Small'}
WORKER_DISK_SIZE=${4:-100}
IMG_NAME=${5:-'Linux Ubuntu 18.04 LTS 64-bit'}

MASTER_VM_PROFILE=Small

DM_VER=v0.16.2
DM_BIN=$HOME/docker-machine

if [ ! -f "$DM_BIN" ]; then
    base=https://github.com/docker/machine/releases/download/$DM_VER
    curl -L $base/docker-machine-"$(uname -s)"-"$(uname -m)" >"$DM_BIN"
fi
chmod +x "$DM_BIN"

SSH_KEY=${SSH_KEY:-${HOME}/.ssh/id_rsa}

if [ ! -f "${SSH_KEY}" ]; then
    echo "creating ${SSH_KEY}"
    yes y | ssh-keygen -q -t rsa -N '' -f ${SSH_KEY} &>/dev/null
fi

#
# A "docker-machine" security group will be created. Specify
# --exoscale-security-group if you need a specific one.
#
BNAME=dockermachine-$(date +%Y%m%d%H%M%S)-k8s
MNAME=${BNAME}-master
CREATE_CMD="create --driver exoscale
    --exoscale-security-group slipstream_managed
    --exoscale-ssh-user root
    --exoscale-ssh-key ${SSH_KEY}
    --exoscale-api-key ${EXOSCALE_API_KEY:?provide EXOSCALE_API_KEY value}
    --exoscale-api-secret-key ${EXOSCALE_API_SECRET:?provide EXOSCALE_API_SECRET value}
    --exoscale-availability-zone ${EXOSCALE_REGION:-CH-GVA-2}"

deploy_master() {
    echo "::: Provisioning master: $MNAME"
    $DM_BIN $CREATE_CMD --exoscale-instance-profile $MASTER_VM_PROFILE \
        --exoscale-image "$IMG_NAME" "$MNAME"
    ip=$($DM_BIN ip "$MNAME")
    $DM_BIN scp ./k8s-install.sh "$MNAME":~
    $DM_BIN ssh "$MNAME" ./k8s-install.sh master
}

deploy_worker() {
    local WNAME=$1
    shift
    local joinToken=$@
    echo "::: Provisioning worker: $WNAME"
    $DM_BIN $CREATE_CMD --exoscale-instance-profile $WORKER_VM_PROFILE \
        --exoscale-disk-size "$WORKER_DISK_SIZE" \
        --exoscale-image "$IMG_NAME" "$WNAME"
    $DM_BIN scp ./k8s-install.sh "$WNAME":~
    $DM_BIN ssh "$WNAME" ./k8s-install.sh worker
    $DM_BIN ssh "$WNAME" "sudo $joinToken"
}

deploy() {
    swarm_node_count=${1:-1}

    deploy_master

    if [ "$swarm_node_count" -ge 1 ]; then
        joinToken=$($DM_BIN ssh "$MNAME" "kubeadm token create --print-join-command" | tr -d '\r')
        for i in $(seq 1 $swarm_node_count); do
            WNAME=${BNAME}-worker${i}
            deploy_worker $WNAME "$joinToken" &
        done
        wait
    fi

    echo "k8s master: $ip"

    $DM_BIN ssh "$MNAME" kubectl get nodes

    config=$(pwd)/config.$BNAME
	echo ::: Copy admin config to $config
    $DM_BIN scp $MNAME:~/.kube/config $config
}

terminate() {
    machines=()
    while read m;do machines+=( "$m" );done < <($DM_BIN ls -q)
    if [ "${#machines[@]}" -eq 0 ]; then
        echo "WARNING: no machines to terminate"
    else
        for m in ${machines[@]};do
            $DM_BIN rm -y "$m" &
        done
        wait
    fi
}

if [ "$ACTION" == "deploy" ]; then
    deploy "$SWARM_NODE_COUNT"
elif [ "$ACTION" == "terminate" ]; then
    terminate
else
    echo "$action_err_msg"
    exit 1
fi


k8s-cluster/cluster-deploy-exoscale.sh
#!/bin/bash

set -xe
set -o errtrace

function cleanup() {
  echo "Running cleanup..."
  $DM_BIN rm -y "$($DM_BIN ls -q)"
  echo "Running cleanup... done."
}

trap cleanup ERR

action_err_msg="Action required: deploy|terminate"
ACTION=${1:?$action_err_msg}
WORKERS_COUNT=${2:-1}
WORKER_VM_PROFILE=${3:-'Medium'}
WORKER_DISK_SIZE=${4:-100}
IMG_NAME=${5:-'Linux Ubuntu 18.04 LTS 64-bit'}

KUBERNETES_VER=1.19.5-00

MASTER_VM_PROFILE=Medium

DM_VER=v0.16.2
DM_BIN=$HOME/docker-machine

if [ ! -f "$DM_BIN" ]; then
    base=https://github.com/docker/machine/releases/download/$DM_VER
    curl -L $base/docker-machine-"$(uname -s)"-"$(uname -m)" >"$DM_BIN"
fi
chmod +x "$DM_BIN"

SSH_KEY=${SSH_KEY:-${HOME}/.ssh/id_rsa}

if [ ! -f "${SSH_KEY}" ]; then
    echo "creating ${SSH_KEY}"
    yes y | ssh-keygen -q -t rsa -N '' -f ${SSH_KEY} &>/dev/null
fi

#
# A "docker-machine" security group will be created. Specify
# --exoscale-security-group if you need a specific one.
#
BNAME=dockermachine-$(date +%Y%m%d%H%M%S)-k8s
MNAME=${BNAME}-master
CREATE_CMD="create --driver exoscale
    --exoscale-security-group slipstream_managed
    --exoscale-ssh-user root
    --exoscale-ssh-key ${SSH_KEY}
    --exoscale-api-key ${EXOSCALE_API_KEY:?provide EXOSCALE_API_KEY value}
    --exoscale-api-secret-key ${EXOSCALE_API_SECRET:?provide EXOSCALE_API_SECRET value}
    --exoscale-availability-zone ${EXOSCALE_REGION:-CH-GVA-2}"

deploy_master() {
    echo "::: Provisioning master: $MNAME"
    $DM_BIN $CREATE_CMD --exoscale-instance-profile $MASTER_VM_PROFILE \
        --exoscale-image "$IMG_NAME" "$MNAME"
    ip=$($DM_BIN ip "$MNAME")
    $DM_BIN scp ./k8s-install.sh "$MNAME":~
    $DM_BIN ssh "$MNAME" ./k8s-install.sh master $KUBERNETES_VER
}

deploy_worker() {
    local WNAME=$1
    shift
    local joinToken=$@
    echo "::: Provisioning worker: $WNAME"
    $DM_BIN $CREATE_CMD --exoscale-instance-profile $WORKER_VM_PROFILE \
        --exoscale-disk-size "$WORKER_DISK_SIZE" \
        --exoscale-image "$IMG_NAME" "$WNAME"
    $DM_BIN scp ./k8s-install.sh "$WNAME":~
    $DM_BIN ssh "$WNAME" ./k8s-install.sh worker $KUBERNETES_VER
    $DM_BIN ssh "$WNAME" "sudo $joinToken"
}

deploy() {
    workers_count=${1:-1}

    deploy_master

    if [ "$workers_count" -ge 1 ]; then
        joinToken=$($DM_BIN ssh "$MNAME" "kubeadm token create --print-join-command" | tr -d '\r')
        for i in $(seq 1 $workers_count); do
            WNAME=${BNAME}-worker${i}
            deploy_worker $WNAME "$joinToken" &
        done
        wait
    fi

    echo "k8s master: $ip"

    $DM_BIN ssh "$MNAME" kubectl get nodes

    config=$(pwd)/kubeconfig.yaml
    echo ::: Copy admin config to $config
    $DM_BIN scp $MNAME:~/.kube/config $config
}

terminate() {
    machines=()
    while read m;do machines+=( "$m" );done < <($DM_BIN ls -q)
    if [ "${#machines[@]}" -eq 0 ]; then
        echo "WARNING: no machines to terminate"
    else
        for m in ${machines[@]};do
            $DM_BIN rm -y "$m" &
        done
        wait
    fi
}

if [ "$ACTION" == "deploy" ]; then
    deploy "$WORKERS_COUNT"
elif [ "$ACTION" == "terminate" ]; then
    terminate
else
    echo "$action_err_msg"
    exit 1
fi
